<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Read sequences from a fasta file. — read_fasta • biozhuoer</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/cerulean/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Read sequences from a fasta file. — read_fasta" />

<meta property="og:description" content="read_fasta() reads all sequences contained in a fasta file
into a tibble (see Value). See FASTA format for the requirements of
input file." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">biozhuoer</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/biozhuoer.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/dongzhuoer/biozhuoer">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Read sequences from a fasta file.</h1>
    <small class="dont-index">Source: <a href='https://github.com/dongzhuoer/biozhuoer/blob/master/R/read-fasta.R'><code>R/read-fasta.R</code></a></small>
    <div class="hidden name"><code>read_fasta.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p><code>read_fasta()</code> reads all sequences contained in a fasta file
into a tibble (see <strong>Value</strong>). See <em>FASTA format</em> for the requirements of
input file.</p>
    
    </div>

    <pre class="usage"><span class='fu'>read_fasta</span>(<span class='no'>file</span>, <span class='kw'>per_line</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>unalign</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>file</th>
      <td><p>string. Path to the fasta file.</p></td>
    </tr>
    <tr>
      <th>per_line</th>
      <td><p>logical scalar. Whether sequences keep in one line or might
span multiple lines. Specify <code>FALSE</code> if not sure, refer to <strong>Details</strong>.</p></td>
    </tr>
    <tr>
      <th>unalign</th>
      <td><p>logical scalar. Whether unalign aligned sequences, i.e. remove
'-', see <strong>Details</strong>.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>tibble</p><ul>
<li><p>name: character. sequence header (without '&gt;')</p></li>
<li><p>seq:  character. sequence itself, I named it <code>seq</code> instead of <code>sequence</code> to
save 62.5<!-- % of time (and everyone using R should take it for granted that seq -->
means sequence)</p></li>
</ul>

    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Previously, I implemented fasta as a named character (in
<code>bioinfor::read_fasta()</code>). But I found it more and more inconvenient as I
used it. Many times when you manipulate the character, the name got lost.
Finally, I decided to reimplement it as a tibble. That why this package
come into being.</p>
<p>Actually, <code>file</code> can be anything as long as it's recognized by
<code><a href='https://readr.tidyverse.org/reference/read_lines.html'>readr::read_lines()</a></code>.</p>
<p>If <code>per_line</code> isn't <code>TRUE</code>, <code>read_fasta()</code> will check it by detecting '&gt;'
and count line number. This (implemented by <code><a href='https://www.rdocumentation.org/packages/stringr/topics/str_detect'>stringr::str_detect()</a></code>) takes
almost the same time as reading file content (implemented by
<code><a href='https://readr.tidyverse.org/reference/read_lines.html'>readr::read_lines()</a></code>). So specify <code>TRUE</code> if you are sure to save time. But
this checking doesn't waste any time when sequences indeed span multiple
lines. Additionly, a file contains odd number of lines will cause
<code>pre_line=TURE</code> to be ignored</p>
<p>For <code>unalign</code>, when deal with large aligned file with linebreak containing
many '-'s, implement <em>unalign</em> (<code><a href='https://www.rdocumentation.org/packages/stringr/topics/str_replace'>stringr::str_replace_all()</a></code> in
<code>read_fasta()</code>) is about 15<!-- % faster than do that externally, i.e. after -->
<code>read_fasta()</code> returns.</p>
    
    <h2 class="hasAnchor" id="fasta-format"><a class="anchor" href="#fasta-format"></a>FASTA format</h2>

    <p>The minimum requirement is that it can only contain
sequences (can use any character except for '&gt;', at least at the begining)
and their headers (begin with '&gt;', one line). Comments are not supported.</p>
<p>Basically, <code><a href='https://www.rdocumentation.org/packages/base/topics/c'>c('&gt;na', 'me', 'ATCG')</a></code> (name span two lines) and <code><a href='https://www.rdocumentation.org/packages/base/topics/c'>c('&gt;name', '&gt;ATCG')</a></code> (sequence begins with '&gt;') are most common (and maybe the only)
errors. Expect that, almost anything is acceptable (though your file may
not be recongnized by other program):</p><ol>
<li><p>name contains '&gt;', see crazy example 1.</p></li>
<li><p>sequence contains '&gt;' (can't be at the begining), see crazy example 2.</p></li>
<li><p>empty name, see crazy example 3.</p></li>
<li><p>empty sequence, see crazy example 4. But it must occupy an empty line if
you want to scape check for <code>pre_line</code> (refer to details).</p></li>
<li><p>very long name, see crazy example 5. It can contain thousands of
thousands characters, so long as you have enough memory (and disk if you
are really mad).</p></li>
<li><p>don't you think the above is enough? We can't make friends anymore.
(Never tell me you have tried crazy example 6)</p></li>
</ol>
    <p>In most situations, you won't have a file like that, unless you
deliberately create one.</p>
    
    <h2 class="hasAnchor" id="test-files"><a class="anchor" href="#test-files"></a>Test files</h2>

    
    <ol>
<li><p>test1.fasta big file with linebreak.</p></li>
<li><p>test12.fasta same file without.</p></li>
<li><p>test2.fasta big aligned file with linebreak.</p></li>
</ol>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.file'>system.file</a></span>(<span class='st'>'extdata'</span>, <span class='st'>'example.fasta'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'biozhuoer'</span>) <span class='kw'>%&gt;%</span> <span class='fu'>read_fasta</span>();</div><div class='output co'>#&gt; <span style='color: #949494;'># A tibble: 62 x 2</span><span>
#&gt;    name                    seq                                                  
#&gt;    </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>                   </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>                                                
#&gt; </span><span style='color: #BCBCBC;'> 1</span><span> Acanthisitta_chloris_J… MVKLGSNLNDKNNKQPSNEDGFQTVPLITPLEVNHLQFPAPEKVIVKTRTEY…
#&gt; </span><span style='color: #BCBCBC;'> 2</span><span> Amazona_aestiva_protein MLAQGKGPQAQCSSQCSSHDASQSHCPPQLPASAGPRHHPAPPHMAPGRRQG…
#&gt; </span><span style='color: #BCBCBC;'> 3</span><span> Anas_platyrhynchos_Jar… MVKLGSNLNDKNSKPPSGEDGFQTVPLITPLEVNHLQFPAPEKVIVKTRTEY…
#&gt; </span><span style='color: #BCBCBC;'> 4</span><span> Anser_cygnoides_protein MCEVSLGVACCVARNLGVKPRKPIRQGKVSQTGQKSSSHYPCDNDRLAGFRS…
#&gt; </span><span style='color: #BCBCBC;'> 5</span><span> Antrostomus_carolinens… QVIVKTRTEYQPDQKNKGKLRVPKIAEFTVSFTDGVTERLKVTILISLALAF…
#&gt; </span><span style='color: #BCBCBC;'> 6</span><span> Apaloderma_vittatum_Ja… MVKLGSNLSDKSSKQPPNEDGFQTVPLITPLEVNHLQFPAPEKVIVKTRTEY…
#&gt; </span><span style='color: #BCBCBC;'> 7</span><span> Aptenodytes_forsteri_J… MVKLGSNLNDKNNKQPSNEDGFQTVPLITPLEVNHLQFPAPEKVIVKTRTEY…
#&gt; </span><span style='color: #BCBCBC;'> 8</span><span> Apteryx_australis_prot… VTILISLALAFLACIVFLVVYKAFTYDHSCPDGFVYKHKRCIPASLDAYYSA…
#&gt; </span><span style='color: #BCBCBC;'> 9</span><span> Aquila_chrysaetos_prot… MVKLGSNLNDKNSKQPSNEDGFQTVPLITPLEVNHLQFPAPEKVIVKTRTEY…
#&gt; </span><span style='color: #BCBCBC;'>10</span><span> Balearica_regulorum_Ja… MVKLGSNLNDKNNKQPSNEDGFQTVPLITPLEVNHLQFPAPEKVIVKTRTEY…
#&gt; </span><span style='color: #949494;'># ... with 52 more rows</span><span></div><div class='input'>
<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.file'>system.file</a></span>(<span class='st'>'extdata'</span>, <span class='st'>'aligned-multiline.fasta'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'biozhuoer'</span>) <span class='kw'>%&gt;%</span> <span class='fu'>read_fasta</span>();</div><div class='output co'>#&gt; </span><span style='color: #949494;'># A tibble: 61 x 2</span><span>
#&gt;    name                    seq                                                  
#&gt;    </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>                   </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>                                                
#&gt; </span><span style='color: #BCBCBC;'> 1</span><span> Acanthisitta_chloris_J… -----------------------------------------------AG---…
#&gt; </span><span style='color: #BCBCBC;'> 2</span><span> Amazona_aestiva_protein MRFWGGGWTLRXT-SEMGNXS----------VQRMRE--------GPGG---…
#&gt; </span><span style='color: #BCBCBC;'> 3</span><span> Anas_platyrhynchos_Jar… ----------------------------------------------RAG---…
#&gt; </span><span style='color: #BCBCBC;'> 4</span><span> Anser_cygnoides_protein ----------------------------------------------------…
#&gt; </span><span style='color: #BCBCBC;'> 5</span><span> Antrostomus_carolinens… -----------------------------------------------AG---…
#&gt; </span><span style='color: #BCBCBC;'> 6</span><span> Apaloderma_vittatum_Ja… ----------------------------------------------RAG---…
#&gt; </span><span style='color: #BCBCBC;'> 7</span><span> Aptenodytes_forsteri_J… -----------------------------------------------AG---…
#&gt; </span><span style='color: #BCBCBC;'> 8</span><span> Apteryx_australis_prot… ----------------------------------------------------…
#&gt; </span><span style='color: #BCBCBC;'> 9</span><span> Aquila_chrysaetos_prot… ---------------------------------------------MFSGCIQ…
#&gt; </span><span style='color: #BCBCBC;'>10</span><span> Balearica_regulorum_Ja… -----------------------------------------------AG---…
#&gt; </span><span style='color: #949494;'># ... with 51 more rows</span><span></div><div class='input'>
<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.file'>system.file</a></span>(<span class='st'>'extdata'</span>, <span class='st'>'aligned-multiline.fasta'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'biozhuoer'</span>) <span class='kw'>%&gt;%</span> <span class='fu'>read_fasta</span>(<span class='kw'>unalign</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>);</div><div class='output co'>#&gt; </span><span style='color: #949494;'># A tibble: 61 x 2</span><span>
#&gt;    name                    seq                                                  
#&gt;    </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>                   </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>                                                
#&gt; </span><span style='color: #BCBCBC;'> 1</span><span> Acanthisitta_chloris_J… AGVLALWALTTHGMYIQDFWRTWLRGLRFFLAVGIFFCVVALVAFCTFLALA…
#&gt; </span><span style='color: #BCBCBC;'> 2</span><span> Amazona_aestiva_protein MRFWGGGWTLRXTSEMGNXSVQRMREGPGGVLALWALITHVMYVQDYWRTWL…
#&gt; </span><span style='color: #BCBCBC;'> 3</span><span> Anas_platyrhynchos_Jar… RAGVLALWALITHVMYVQDYWRTWLKGLRFFLFIGILFSALSVVGICTFLVL…
#&gt; </span><span style='color: #BCBCBC;'> 4</span><span> Anser_cygnoides_protein MYVQDYWRTWLKGLRFFLFIGILFSALSVVGFCTFLVLAITKHQSLTDPRSY…
#&gt; </span><span style='color: #BCBCBC;'> 5</span><span> Antrostomus_carolinens… AGVLALWALITHVMYVQDYWRTWLKGLRFFLFIGILFSALSVVGFCTFLVLA…
#&gt; </span><span style='color: #BCBCBC;'> 6</span><span> Apaloderma_vittatum_Ja… RAGVLALWALTTHVMYVQDYWRTWLKGLRFFLFIGILFSALSAVGFCAFLVL…
#&gt; </span><span style='color: #BCBCBC;'> 7</span><span> Aptenodytes_forsteri_J… AGVLALWALITHVMYVQDYWRTWLKGLRFFLFIGILFSALSVVGFCTFLVLA…
#&gt; </span><span style='color: #BCBCBC;'> 8</span><span> Apteryx_australis_prot… MHLSFPLESDYWRTWLKGLRFFLFIGILFSALSVVGFCTFLVLAITKHQSLT…
#&gt; </span><span style='color: #BCBCBC;'> 9</span><span> Aquila_chrysaetos_prot… MFSGCIQPFPEWLLMFLCVLALWALITHVMYVQDYWRTWLKGLRFFLFIGIL…
#&gt; </span><span style='color: #BCBCBC;'>10</span><span> Balearica_regulorum_Ja… AGVLALWALITHVMYVQDYWRTWLKGLRFFLFIGILFSALSVVGFCTFLVLA…
#&gt; </span><span style='color: #949494;'># ... with 51 more rows</span><span></div><div class='input'>

<span class='co'># crazy examples</span>
<span class='fu'>read_fasta</span>(<span class='st'>'&gt;na&gt;me\nATCG'</span>);</div><div class='output co'>#&gt; </span><span style='color: #949494;'># A tibble: 1 x 2</span><span>
#&gt;   name  seq  
#&gt;   </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span> </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>
#&gt; </span><span style='color: #BCBCBC;'>1</span><span> na&gt;me ATCG </div><div class='input'>
<span class='fu'>read_fasta</span>(<span class='st'>'&gt;name\nAT&gt;CG'</span>);</div><div class='output co'>#&gt; </span><span style='color: #949494;'># A tibble: 1 x 2</span><span>
#&gt;   name  seq  
#&gt;   </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span> </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>
#&gt; </span><span style='color: #BCBCBC;'>1</span><span> name  AT&gt;CG</div><div class='input'>
<span class='fu'>read_fasta</span>(<span class='st'>'&gt;\nATCG'</span>);</div><div class='output co'>#&gt; </span><span style='color: #949494;'># A tibble: 1 x 2</span><span>
#&gt;   name  seq  
#&gt;   </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span> </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>
#&gt; </span><span style='color: #BCBCBC;'>1</span><span> </span><span style='color: #949494;'>""</span><span>    ATCG </div><div class='input'>
<span class='fu'>read_fasta</span>(<span class='st'>'&gt;name\n'</span>)</div><div class='output co'>#&gt; </span><span style='color: #949494;'># A tibble: 1 x 2</span><span>
#&gt;   name  seq  
#&gt;   </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span> </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>
#&gt; </span><span style='color: #BCBCBC;'>1</span><span> name  </span><span style='color: #949494;'>""</span><span>   </div><div class='input'>
<span class='fu'>read_fasta</span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/paste'>paste0</a></span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>'&gt;'</span>, <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/rep'>rep</a></span>(<span class='st'>'x'</span>, <span class='fl'>10000</span>), <span class='st'>'\nATCG'</span>), <span class='kw'>collapse</span> <span class='kw'>=</span> <span class='st'>''</span>));</div><div class='output co'>#&gt; </span><span style='color: #949494;'># A tibble: 1 x 2</span><span>
#&gt;   name                                                                     seq  
#&gt;   </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>                                                                    </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>
#&gt; </span><span style='color: #BCBCBC;'>1</span><span> xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx… ATCG </div><div class='input'>
<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/tempfile'>tempfile</a></span>() <span class='kw'>%T&gt;%</span> <span class='kw pkg'>readr</span><span class='kw ns'>::</span><span class='fu'><a href='https://readr.tidyverse.org/reference/read_file.html'>write_file</a></span>(<span class='st'>'&gt;name'</span>, <span class='no'>.</span>)  <span class='kw'>%T&gt;%</span> <span class='kw pkg'>readr</span><span class='kw ns'>::</span><span class='fu'><a href='https://readr.tidyverse.org/reference/read_file.html'>read_file</a></span>() <span class='kw'>%&gt;%</span> <span class='no'>read_fasta</span>;</div><div class='output co'>#&gt; </span><span style='color: #949494;'># A tibble: 1 x 2</span><span>
#&gt;   name  seq  
#&gt;   </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span> </span><span style='color: #949494;font-style: italic;'>&lt;chr&gt;</span><span>
#&gt; </span><span style='color: #BCBCBC;'>1</span><span> name  </span><span style='color: #949494;'>""</span><span>   </div><div class='input'>

</div></span></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#fasta-format">FASTA format</a></li>

      <li><a href="#test-files">Test files</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://dongzhuoer.com'>Zhuoer Dong</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.9000.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

